sound/soc/intel/boards/bdw-rt5677.c

```diff
@@ -307,6 +307,15 @@ static int bdw_rt5677_init(struct snd_soc_pcm_runtime *rtd)
 	return 0;
 }
 
+static void bdw_rt5677_exit(struct snd_soc_pcm_runtime *rtd)
+{
+	struct bdw_rt5677_priv *bdw_rt5677 =
+			snd_soc_card_get_drvdata(rtd->card);
+
+	if (!IS_ERR(bdw_rt5677->gpio_hp_en))
```

Comment:
can this happen? If `gpiod_get()` above fails and returns an error code, the initialisation fails, so `.exit()` will never be called?

Reply:
Good feedback, but I think it could happen that the exit() is reached with invalid data.

look at the code in sound/soc/soc-core.c. Whether you have an error in soc_init_pcm_runtime() - which calls snd_soc_link_init() - or later, you jump to the probe_end label, so the flow is this

snd_soc_bind_card()
    (snd_soc_link_init() may or may not be called, and may or may not be succesful)
    probe_end
        soc_cleanup_card_resources(card);
             snd_soc_remove_pcm_runtime(card, rtd);
                  snd_soc_link_exit(rtd); <<< KABOOM

So in exit() you *have* to check if the resources you are about to release where actually allocated.

The test should actually be IS_ERR_OR_NULL() to cover all cases.